packages/gatsby/src/services/listen-to-webpack.ts

```diff
@@ -7,6 +7,20 @@ export const createWebpackWatcher = (compiler: Compiler): InvokeCallback => (
 ): void => {
   compiler.hooks.invalid.tap(`file invalidation`, file => {
     reporter.verbose(`Webpack file changed: ${file}`)
-    callback({ type: `SOURCE_FILE_CHANGED`, file })
+
+    // Webpack fires `invalid` event as soon as any file changes
+    //   but it doesn't start recompiling at this point. Instead, it aggregates changes with debounce
+    //   and recompile on a tail of debounce.
+    //   For example, you may save a file multiple times quickly but webpack will only recompile once.
+    //   Unfortunately webpack doesn't expose any event for aggregated changes.
+    //   But we actually need it. If we start recompiling immediately on `invalid` we can miss some changes.
+    // Hack below is a workaround for this missing webpack event
+    // @ts-ignore
+    const watcher = compiler.watchFileSystem.watcher // Watchpack instance
+    watcher.once(`aggregated`, () => {
+      callback({ type: `SOURCE_FILE_CHANGED`, file })
+    })
+
+    // TODO: fallback to timeout?
```

Comment:
from my own investigations - sometimes webpack/watchpack can emit `aggregated` without emitting `invalid` before - should this maybe be hoisted and just listening to `watcher.on('aggregated')` and just avoid using `hooks.invalid`?

Reply:
It is tricky because each compiliation has a new `compiler.watchFileSystem.watcher` instance. And we want to listen to current active instance. But also - if it doesn't emit `invalid` - how do we even react to this now? Does it mean we don't emit `SOURCE_FILE_CHANGED` at all? Or we do but in query-watcher?